### DISCUSSION TITLE
Config Watcher for SQL database instead of .yaml configuration file

### DISCUSSION THREAD

**Discussion by jasonmccomb** [bluenviron/mediamtx#1435] [General]
_I asked ChatGPT for help writing a Docker Compose file - it replied with my parameters:_

"Sure, here is an example docker-compose.yml file that uses the aler9/rtsp-simple-server image, mounts a volume containing an rtsp-simple-server.yml file, and also starts an SQL database service using the mariadb image:

`
version: '3'
services:
  rtsp-server:
    image: aler9/rtsp-simple-server
    volumes:
      - "./rtsp-simple-server.yml:/etc/rtsp-simple-server/rtsp-simple-server.yml"
    ports:
      - "8554:8554"
  db:
    image: mariadb
    volumes:
      - "./db:/var/lib/mysql"
    environment:
      MYSQL_ROOT_PASSWORD: password
      MYSQL_DATABASE: rtsp-db
      MYSQL_USER: rtsp-user
      MYSQL_PASSWORD: rtsp-password
    ports:
      - "3306:3306"
`

This file creates a service called "rtsp-server" that runs the aler9/rtsp-simple-server image and maps port 8554 on the host to port 8554 on the container. It also mounts a volume containing the rtsp-simple-server.yml file to /etc/rtsp-simple-server/rtsp-simple-server.yml inside the container, replacing the default configuration file.

It also creates a service called "db" that runs the mariadb image, maps port 3306 on the host to port 3306 on the container, and sets environment variables for the root password, the database name, the user and the password. The db service also mounts a volume in the host to /var/lib/mysql in the container, so the data can be persistent.

You can then use the docker-compose up -d command to start the services in detached mode, and docker-compose down to stop them.

It's important to notice that you need to have the rtsp-simple-server.yml file in the same directory as the docker-compose.yml file, and the db folder inside the directory too.

_**Also, it's important to be aware that if you change the rtsp-simple-server.yml file, you need to stop and start the container for the changes to take effect.**_ (emphasis added)

Please note that I am using mariadb as an example, you should use the database of your choice and configure the environment variables accordingly.

You will also need to create the fields in the database that corresponds to the rtsp-simple-server.yml file, and make sure the application is correctly configured to connect and use the database. "

_My next question:_

if the sql database is used to update the configuration file in the container, would it need to be stopped and restarted for the server to update its behavior?

"It depends on how the application that runs in the container is designed to update its configuration. If the application is designed to read the configuration from the SQL database, then the application should be able to update its behavior without the need to stop and restart the container.

However, if the application is designed to read the configuration from a file, and the file is updated by writing the new configuration to the file from the database, then the container would need to be stopped and restarted for the application to read the new configuration from the file.

Another way of updating the application behavior, would be to have the application to have a way of listening for changes in the database and updating itself accordingly, but this would need to be implemented in the application and it would be dependent of the application design.

It's important to note that the example I provided earlier is just a basic example, you would need to make sure the application is designed to work with the database and that the fields in the database corresponds to the fields in the rtsp-simple-server.yml file, otherwise the application behavior might not change as expected."

_My next question:_

This is the code for the configuration file change watcher: rtsp-simple-server/internal/confwatcher/confwatcher.go. Could you suggest better code that would monitor the sql database for configuration changes?

"I am not aware of any specific code that monitors an SQL database for configuration changes in the rtsp-simple-server package, but here is an example of how you could implement a configuration change watcher that monitors an SQL database for changes:

`package confwatcher

import (
    "database/sql"
    "time"
)

func WatchSQL(db *sql.DB, interval time.Duration, callback func()) {
    for {
        // Execute a query to check for changes in the configuration table
        var count int
        err := db.QueryRow("SELECT COUNT(*) FROM configuration WHERE updated_at > ?", lastUpdate).Scan(&count)
        if err != nil {
            log.Printf("Error checking for configuration updates: %v", err)
        } else if count > 0 {
            // Configuration has changed, call the callback function
            callback()
            // Update the lastUpdate timestamp
            lastUpdate = time.Now()
        }
        // Sleep for the specified interval before checking again
        time.Sleep(interval)
    }
}
`

This code defines a WatchSQL function that takes a pointer to an SQL database connection, an interval duration, and a callback function as arguments. The function runs in a loop and executes a query to check for changes in the configuration table. If the query returns any rows, it means that the configuration has changed, and it calls the callback function.

You would need to replace configuration and updated_at with the actual table name and the field name that indicates when the configuration was last updated in your database.

Additionally, you would need to pass an instance of the database driver you are using (e.g sql.DB from the database/sql package) and the callback function that you want to be called when a change is detected.

It's important to notice that this is a simple example, you should take into account the performance and scalability of your solution, especially if you expect a high number of configuration changes.

Also, make sure to handle the errors and logging as necessary for your application."

### how correct is ChatGPT?

I think the ability to update the server configuration on the fly is an excellent feature of rtsp–simple–server. I'm going to test the assertion that a container restart is required for the changes to take effect. 

If the 'sql config watcher' idea works, using multiple Docker containers to run instances of rtsp-simple-server seems manageable. Common media assets would reside on a shared volume. The central database manages each stream in real time along with all the other database functions for media services. 

Wowza without the cost? 

I would really enjoy hearing from anyone on this. Thank you.

PS -- as I continue learning about simple server, and for running a general, I see that multiple streams are a feature of the application.

**Comment by JeDiE99** [comment]
see this https://github.com/aler9/rtsp-simple-server/issues/1406
relational DB is not good for observing config

**Reply by jasonmccomb** [reply]
Thank you I'll take a look

**Comment by aler9** [comment]
I understand the hype for ChatGPT, but currently its answers are mostly wrong. This server doesn't make use of MySQL. Furthermore, the suggested watching techique, which is called polling, is pretty inefficient. If you want to generate a docker-compose file for the server, just take the suggested Docker command line in the README and start from that.

**Comment by aler9** [comment]
Another thing - Wowza is a commercial product which is much more versatile and has a different user base.

This is an open source project focused on real time and efficiency, with much less features than Wowza, but compatible with both Raspberry Pis and enterprise servers. No one here wants to clone or replace Wowza, there is no man power nor the will to do so.

[a comment by an automated account is not shown]
